### PR TITLE
chore(awslint): add rule to disallow onPrepare methods

### PR DESCRIPTION
**Pull request by Chriscbr**
The `onPrepare` lifecycle hook is deprecated in CDK v1, so it causes conflicts when forward merging code to CDK v2. Add an awslint rule to enforce this.

Unfortunately, because of limitations in `awslint` the rule only ends up getting applied to public APIs, so it doesn't catch `onPrepare` usages on private classes, but this is still an improvement.

----

### All Submissions:

* [ ] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/integ-tests/lib/assertions/providers/provider.ts [hunk 1]
```diff
@@ -115,13 +115,16 @@ class SingletonFunction extends CoreConstruct {
...
-    this.lambdaFunction.addPolicies(this.policies);
+    /**
+     * The policies can be added by different constructs
+     */
+    this.node.addValidation({
+      validate: () => {
+        this.lambdaFunction.addPolicies(this.policies);
+        return [];
+      },
+    });
```
Comment (rix0rrr) [inline packages/@aws-cdk/integ-tests/lib/assertions/providers/provider.ts:127]: This needs to be an `Aspect`, not a validation.

File: packages/@aws-cdk/integ-tests/lib/test-case.ts [hunk 2]
```diff
@@ -146,14 +146,16 @@ export class IntegTest extends CoreConstruct {
...
-        synthesizer.synthesize(session);
+    this.node.addValidation({
+      validate: () => {
+        attachCustomSynthesis(this, {
+          onSynthesize: (session: ISynthesisSession) => {
+            const synthesizer = new IntegManifestSynthesizer(this.testCases);
+            synthesizer.synthesize(session);
+          },
+        });
+        return [];
```
Comment (rix0rrr) [inline packages/@aws-cdk/integ-tests/lib/test-case.ts:158]: Same.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/integ-tests/lib/assertions/providers/provider.ts (hunk 1 above)
Comment (Chriscbr) [reply packages/@aws-cdk/integ-tests/lib/assertions/providers/provider.ts:127]: (for my own learning) What's the difference?

**Comment by Chriscbr** [comment]
I'm not super familiar onPrepare or Aspects but the tests passed on my machine so I _think_ it's working.

Pinging @corymhall to make sure I haven't done anything bad with this integ testing code 🙂

**Comment by corymhall** [comment]
> I'm not super familiar onPrepare or Aspects but the tests passed on my machine so I _think_ it's working.
> 
> Pinging @corymhall to make sure I haven't done anything bad with this integ testing code 🙂

looks good to me!

File: packages/@aws-cdk/integ-tests/lib/assertions/providers/provider.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/integ-tests/lib/assertions/providers/provider.ts]: It's when they run and what they're expected to do.

Aspects are allowed to change the construct tree, Validations are not.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/integ-tests/lib/assertions/providers/provider.ts (hunk 1 above)
Comment (Chriscbr) [reply packages/@aws-cdk/integ-tests/lib/assertions/providers/provider.ts]: I see, that makes sense - thanks!

**Pull request closed without merging**